From a9e0dd30379148d7d232ed27056ab36b206f96a6 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Thu, 30 Jan 2025 14:52:59 +0100
Subject: [PATCH] phy: stm32: usb2phy: enable wakeup source, report wakeup
 events

Enable wakeup capability on stm32-usb2phy, and report wakeup events.
Adopt the pm_wakeup_hard_event(), in order to prevent race condition
while suspending, e.g. in case the host resumes the bus.

Change-Id: I90f4f02d2f895b8203fefc32d1f9b8192689151f
Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/433272
Reviewed-by: Pankaj DEV <pankaj.dev@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/phy/st/phy-stm32-usb2phy.c | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

--- a/drivers/phy/st/phy-stm32-usb2phy.c
+++ b/drivers/phy/st/phy-stm32-usb2phy.c
@@ -527,9 +527,12 @@ static int stm32_usb2phy_phy_power_on(st
 	struct stm32_usb2phy *phy_dev = phy_get_drvdata(phy);
 	struct device *dev = &phy->dev;
 
-	if (phy_dev->wakeirq > 0)
+	if (phy_dev->wakeirq > 0) {
 		if (enable_irq_wake(phy_dev->wakeirq))
 			dev_warn(dev, "Wake irq not enabled\n");
+		if (device_wakeup_enable(phy_dev->dev))
+			dev_warn(dev, "device_wakeup_enable failed\n");
+	}
 
 	if (phy_dev->vbus)
 		return regulator_enable(phy_dev->vbus);
@@ -542,9 +545,12 @@ static int stm32_usb2phy_phy_power_off(s
 	struct stm32_usb2phy *phy_dev = phy_get_drvdata(phy);
 	struct device *dev = &phy->dev;
 
-	if (phy_dev->wakeirq > 0)
+	if (phy_dev->wakeirq > 0) {
+		if (device_wakeup_disable(phy_dev->dev))
+			dev_warn(dev, "device_wakeup_disable failed\n");
 		if (disable_irq_wake(phy_dev->wakeirq))
 			dev_warn(dev, "Wake irq not disabled\n");
+	}
 
 	if (phy_dev->vbus)
 		return regulator_disable(phy_dev->vbus);
@@ -782,6 +788,11 @@ static int stm32_usb2phy_tuning(struct p
 
 static irqreturn_t stm32_usb2phy_irq_wakeup_handler(int irq, void *dev_id)
 {
+	struct device *dev = dev_id;
+
+	/* Prevents remote wakeup interrupt race while suspending */
+	pm_wakeup_hard_event(dev);
+
 	return IRQ_HANDLED;
 }
 
@@ -841,10 +852,12 @@ static int stm32_usb2phy_probe(struct pl
 
 		ret = devm_request_threaded_irq(dev, phy_dev->wakeirq, NULL,
 						stm32_usb2phy_irq_wakeup_handler, IRQF_ONESHOT,
-						NULL, NULL);
+						NULL, dev);
 		if (ret)
 			return dev_err_probe(dev, ret, "unable to request wake IRQ %d\n",
 						 phy_dev->wakeirq);
+
+		device_set_wakeup_capable(dev, true);
 	}
 
 	phy_dev->hw_data = stm32_usb2phy_get_hwdata(dev, phycr);
